pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -220,11 +255,19 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
   protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
```

Comment:
Nice job handling the previous record deleted.

When the new record is a delete record, we can skip reading the previous record and updating the new record. We do want to still handle the out-of-order case. We can consider adding a flag for `outOfOrderEvent` to replace the current `previousRecordDeleted` flag

Reply:
> When the new record is a delete record, we can skip reading the previous record and updating the new record. 

Ah, I think it should still merge with previous record because technically, you could view it as just another update to the delete column.

Reply:
I don't follow. When it gets a delete record, the old record is considered deleted, and all other column values shouldn't matter anymore.